// @volar/vue-typescript/out/plugins/vue-sfc-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SourceMaps = require("@volar/source-map");
const sourceMaps_1 = require("../utils/sourceMaps");
function default_1() {
    return {
        getEmbeddedFilesCount(sfc) {
            return sfc.template ? 1 : 0;
        },
        getEmbeddedFile(fileName, sfc, i) {
            const template = sfc.template;
            const file = {
                fileName: fileName + '.' + template.lang,
                lang: template.lang,
                content: template.content,
                capabilities: {
                    diagnostics: true,
                    foldingRanges: true,
                    formatting: true,
                    documentSymbol: true,
                    codeActions: true,
                    inlayHints: true,
                },
                isTsHostFile: false,
            };
            const sourceMap = new sourceMaps_1.EmbeddedFileSourceMap();
            sourceMap.mappings.push({
                data: {
                    vueTag: template.tag,
                    capabilities: {
                        basic: true,
                        references: true,
                        definitions: true,
                        diagnostic: true,
                        rename: true,
                        completion: true,
                        semanticTokens: true,
                    },
                },
                mode: SourceMaps.Mode.Offset,
                sourceRange: {
                    start: 0,
                    end: template.content.length,
                },
                mappedRange: {
                    start: 0,
                    end: template.content.length,
                },
            });
            const embedded = {
                file,
                sourceMap,
            };
            return embedded;
        },
    };
}
exports.default = default_1;
